DQM: Sequence inspection tool and test

schneiml opened the pull request:
#### PR description:

This PR introduces a new tool that instruments `cmsDriver/cmsRun` to find out which modules are effectively loaded for each sequence. It is not strictly limited to DQM sequences, but focused on this usage for now. The method used is rather complicated and fragile, but it gives accurate results, unlike other methods I tried.

The information gathered about all sequences mentioned in `runTheMatrix` is stored into a SQLite database that can then be browsed interactively using a simple web-based frontend, or queried automatically.

This PR includes a unit test that builds the database and then queries for DQM legacy modules, and fails if any are found (it currently finds but ignores legacy `EDProducer`s and `EDFilter`s). The test runs multi-threaded and still takes about 10-20min to complete; I am not sure if this is ok for a unit test.

The tool can also be used to display the contents of a single sequence (see `--help`).

The tool works by
- Querying `runTheMatrix.py` for parameter combinations passed to `cmsDriver.py`.
- Executing `cmsDriver.py` once for each parameter combination, along with some fake options that should not affect which modules run but let `cmsDriver.py` succeed.
- Executing the resulting config files with `cmsRun` with the tracer service enabled. The jobs are configured for 0 events, so they finish rather quickly without actually reading any input, except for the HARVESTING jobs which tend to do things in `endJob` and eventually crash. This is fine though, the main purpose of actually running `cmsRun` is to get the list of loaded modules out of the tracer output.
- Dumping the configuration using `edmConfigDump` and parsing the output to resolve the module labels from the tracer into class names, and collect the precise config fragment for each instance.
- Running `edmPluginHelp` for each class to see what plugin type it is.

`cmsDriver` and `edmPluginHelp` are pretty slow, things are done multi-threaded to finish in reasonable time. If anybody has suggestions on how to avoid this huge detour and still produce reliable results, input is appreciated.

I may provide a an instance of the web interface hosted somewhere at some point, but for now, you need to run it yourself if you want to try it (Edit: test setup here: http://cmsswconfigexplore.cern.ch:8000/ (CERN network only), no sure for how long ti will keep running):

```
firefox http://localhost:8000 &
ssh -L 8000:localhost:8000 lxplus.cern.ch
mkdir -p /tmp/$USER
cd /tmp/$USER
cmsrel CMSSW_11_1_0_pre4
cd CMSSW_11_1_0_pre4/src
cmsenv
git cms-merge-topic schneiml:dqm-dqmdumpsequence
scram b
cmsswSequenceInfo.py --runTheMatrix
cmsswSequenceInfo.py --serve
```

#### PR validation:
This PR should not affect any production workflows. The tool seem to work.

The unit test might be very sensitive to all sorts of configuration problems.

schneiml committed:
A simple tool to show what a DQM sequence expands to.

Running cmsDriver and even cmsRun is quite a detour, but at least I have some confidence that the results are realistic.

schneiml committed:
Make dqmdumpsequence.sh more of a real tool.

schneiml committed:
Merge some logic from edmCheckMultithreading into dqmdumpsequence.

schneiml committed:
Move dqmdumpsequence to Configuration, fits better there.

schneiml committed:
Introduce a proper tool to analyze the content of sequences, that scale to checking all.

schneiml committed:
Handle more cmsDriver options and properly implement output.

schneiml committed:
Also track workflows.

schneiml committed:
Implement web frontend.

schneiml committed:
Some more usability improvements.

schneiml committed:
Some cleanup & comments.

schneiml committed:
Fix whitespace.

schneiml committed:
cmsswSequenceInfo.py should by strictly superior now.

schneiml committed:
Unit test using cmsswSequenceInfo.

Dumps all the DQM and VALIDATION sequences and fails if there is a
legacy EDAnalyzer on them (reporting the workflow and sequence to the
log in case).

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-29150/14084

- This PR adds an extra 32KB to repository

- Found files with invalid states:
  - DQMServices/Components/scripts/dqmdumpsequence.sh:
    - Added: 81093b2993dabe5f2b0f4c127d32d1c71adfc5a4
    - Modified: f3607ea66f862dae77266ede2bea80344c3a82e9, 5c4319ed75ade7f6846e267fa17c1432c7e5ee9b
    - Deleted: 85eb5b3f1f8a4bfe9eaca18b2fff5db78d1c3a81
  - DQMOffline/Configuration/scripts/dqmdumpsequence.sh:
    - Added: 85eb5b3f1f8a4bfe9eaca18b2fff5db78d1c3a81
    - Deleted: c032cecc42d861e746c98247fd66b4475bc8c0d8

cmsbuild commented:
A new Pull Request was created by  @schneiml (Marcel Schneider) for master.

It involves the following packages:

DQMOffline/Configuration


@andrius-k, @kmaeshima, @schneiml, @cmsbuild, @jfernan2, @fioriNTU can you please review it and eventually sign? Thanks.
@threus, @rociovilar this is something you requested to watch as well.
@davidlange6, @silviodonato, @fabiocos you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

schneiml committed:
Clarify runtests.sh.

Apparently the ! operator might not work with set -e.

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-29150/14085

- This PR adds an extra 32KB to repository

- Found files with invalid states:
  - DQMServices/Components/scripts/dqmdumpsequence.sh:
    - Added: 81093b2993dabe5f2b0f4c127d32d1c71adfc5a4
    - Modified: f3607ea66f862dae77266ede2bea80344c3a82e9, 5c4319ed75ade7f6846e267fa17c1432c7e5ee9b
    - Deleted: 85eb5b3f1f8a4bfe9eaca18b2fff5db78d1c3a81
  - DQMOffline/Configuration/scripts/dqmdumpsequence.sh:
    - Added: 85eb5b3f1f8a4bfe9eaca18b2fff5db78d1c3a81
    - Deleted: c032cecc42d861e746c98247fd66b4475bc8c0d8

cmsbuild commented:
Pull request #29150 was updated. @andrius-k, @kmaeshima, @schneiml, @cmsbuild, @jfernan2, @fioriNTU can you please check and sign again.

schneiml commented:
please test

I assume this would run the new unit test? Let's see how it goes.

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: e7855caf71bd22b97e68c771340db3073b6fe277
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-28d37e/5080/summary.html
CMSSW: CMSSW_11_1_X_2020-03-09-1100
SCRAM_ARCH: slc7_amd64_gcc820

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-28d37e/5080/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 34
 - DQMHistoTests: Total histograms compared: 2680577
 - DQMHistoTests: Total failures: 39
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2680219
 - DQMHistoTests: Total skipped: 319
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 33 files compared)
 - Checked 147 log files, 16 edm output root files, 34 DQM output files

schneiml commented on DQMOffline/Configuration/scripts/cmsswSequenceInfo.py:
```diff
@@ -0,0 +1,532 @@
+
+# We use two global thread pools, to avoid submitting from one Pool into itself.
+tp = ThreadPool()
+stp = ThreadPool()
```
@smuzaffar @silviodonato We could make the number of threads configurable here quite easily for the unit tests.

smuzaffar commented on DQMOffline/Configuration/test/BuildFile.xml:
```diff
@@ -0,0 +1,5 @@
+<use   name="FWCore/Framework"/>
+<use   name="FWCore/Utilities"/>
+<bin file="TestRunner.cpp" name="TestDQMOfflineConfiguration">
+  <flags TEST_RUNNER_ARGS=" /bin/bash DQMOffline/Configuration/test runtests.sh"/>
```
you can delete DQMOffline/Configuration/test/TestRunner.cpp and use the following to run the test. 
```
<test name="TestDQMOfflineConfiguration" command="runtests.sh --threads 1"/>
```
add `--threads` option to  cmsswSequenceInfo.py so that scram runtests always use 1 thread. Or if possible to break this tests in many then you can do something like

```
<test name="TestDQMOfflineConfiguration1" command="runtests.sh --threads 1 --workflows 1,2,3"/>
<test name="TestDQMOfflineConfiguration2" command="runtests.sh --threads 1 --workflows 4,5,6"/>
<test name="TestDQMOfflineConfiguration3" command="runtests.sh --threads 1 --workflows 7,8"/>
<test name="TestDQMOfflineConfiguration" command="runtests-final.sh">
  <flags PRE_TEST="TestDQMOfflineConfiguration1"/>
  <flags PRE_TEST="TestDQMOfflineConfiguration2"/>
  <flags PRE_TEST="TestDQMOfflineConfiguration3"/>
</test>
```
in this case scram can run TestDQMOfflineConfiguration1,2,3 in parallel.

schneiml replied:
@smuzaffar thanks for the hint, I copied that pattern from other tests... (which probably also should not use it)

It should be pretty easy to break the test down into `DQM` and `VALIDATION`, more would get messy. I'll try that tomorrow.

schneiml committed:
Adjust the tests to run single-threaded

... but in two parts. This should be better for the IB system. Also use
the correct BuildFile commands, thanks @smuzaffar!

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-29150/14149

- This PR adds an extra 36KB to repository

- Found files with invalid states:
  - DQMServices/Components/scripts/dqmdumpsequence.sh:
    - Added: 81093b2993dabe5f2b0f4c127d32d1c71adfc5a4
    - Modified: f3607ea66f862dae77266ede2bea80344c3a82e9, 5c4319ed75ade7f6846e267fa17c1432c7e5ee9b
    - Deleted: 85eb5b3f1f8a4bfe9eaca18b2fff5db78d1c3a81
  - DQMOffline/Configuration/scripts/dqmdumpsequence.sh:
    - Added: 85eb5b3f1f8a4bfe9eaca18b2fff5db78d1c3a81
    - Deleted: c032cecc42d861e746c98247fd66b4475bc8c0d8
  - DQMOffline/Configuration/test/TestRunner.cpp:
    - Added: ab0f9700ecdcecac457d92ba8739ce5bd8a7f125
    - Deleted: 8f6d3f7fbc1416b979685724e7b599881638140c

cmsbuild commented:
Pull request #29150 was updated. @andrius-k, @kmaeshima, @schneiml, @cmsbuild, @jfernan2, @fioriNTU can you please check and sign again.

schneiml commented:
please test

I split the test in two pieces and run each single threaded now. I have not run that to completion yet, let's see how it goes.

cmsbuild commented:
The tests are being triggered in jenkins.

smuzaffar commented:
please test

cmsbuild commented:
-1

Tested at: 8f6d3f7fbc1416b979685724e7b599881638140c

CMSSW: CMSSW_11_1_X_2020-03-10-2300
SCRAM_ARCH: slc7_amd64_gcc820
You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-28d37e/5130/summary.html

I found follow errors while testing this PR

**Failed tests:  UnitTests**

* **Unit Tests**:

I found errors in the following unit tests: 
 
 ---> test TestDQMOfflineConfigurationDQM had ERRORS

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-28d37e/5130/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 34
 - DQMHistoTests: Total histograms compared: 2680577
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2680257
 - DQMHistoTests: Total skipped: 319
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 33 files compared)
 - Checked 147 log files, 16 edm output root files, 34 DQM output files

schneiml commented:
It looks like the test timed out? Also it appears that the division into DQM and VALIDATION is not very even and DQM is a much bigger chunk. 

I guess then the next option is to actually divide the sequences onto multiple tests...

smuzaffar commented:
@schneiml , how about something like 

```
<test name="DumpDQMOfflineConfigurationWorkflows" command="runtests-dump.sh N" />
```
which can find all the workflows and divide them in N parts (write in to some files with N lines in it)

and
```
<test name="TestDQMOfflineConfiguration1" command="runtests.sh 1">
  <flags PRE_TEST="DumpDQMOfflineConfigurationWorkflows"/>
</test>
<test name="TestDQMOfflineConfiguration2" command="runtests.sh 2">
  <flags PRE_TEST="DumpDQMOfflineConfigurationWorkflows"/>
</test>
<test name="TestDQMOfflineConfiguration3" command="runtests.sh 3">
  <flags PRE_TEST="DumpDQMOfflineConfigurationWorkflows"/>
</test>
.....
<test name="TestDQMOfflineConfigurationN" command="runtests.sh N">
  <flags PRE_TEST="DumpDQMOfflineConfigurationWorkflows"/>
</test>
``` 
where runtests.sh  reads the Nth line to get the workflow number and process it. If runtests.sh finds an empty line then just exit 0

schneiml commented:
@smuzaffar is there any reason to use the `PRE_TEST`? I can just run all the batches completely independently (just test-running it). The only issue is that new batches need to be added if the number of sequences to be checked increases, but I added a test that will fail if there are more sequences than are covered by the batches, so that should be fine.

smuzaffar commented:
@schneiml , PRE_TEST is needed to make sure that DumpDQMOfflineConfigurationWorkflows runs before running any TestDQMOfflineConfigurationN

schneiml committed:
Divide work into smaller chunks.

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
The code-checks are being triggered in jenkins.

schneiml commented:
This version now does 5 chunks of up to 50 sequences each, and finished in 1h on `cmsdev32`. We could go even smaller, but there is increasing overhead since each job needs to run `edmPluginHelp` for ~all modules (inside the job it can be cached, but across jobs that would be complicated).

Now, the hard part is to have confidence that this will still actually trigger if there are legacy modules, but by construction it should be good.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-29150/14170

- This PR adds an extra 56KB to repository

- Found files with invalid states:
  - DQMServices/Components/scripts/dqmdumpsequence.sh:
    - Added: 81093b2993dabe5f2b0f4c127d32d1c71adfc5a4
    - Modified: f3607ea66f862dae77266ede2bea80344c3a82e9, 5c4319ed75ade7f6846e267fa17c1432c7e5ee9b
    - Deleted: 85eb5b3f1f8a4bfe9eaca18b2fff5db78d1c3a81
  - DQMOffline/Configuration/scripts/dqmdumpsequence.sh:
    - Added: 85eb5b3f1f8a4bfe9eaca18b2fff5db78d1c3a81
    - Deleted: c032cecc42d861e746c98247fd66b4475bc8c0d8
  - DQMOffline/Configuration/test/TestRunner.cpp:
    - Added: ab0f9700ecdcecac457d92ba8739ce5bd8a7f125
    - Deleted: 8f6d3f7fbc1416b979685724e7b599881638140c

cmsbuild commented:
Pull request #29150 was updated. @andrius-k, @kmaeshima, @schneiml, @cmsbuild, @jfernan2, @fioriNTU can you please check and sign again.

schneiml commented:
please test

Now the new test should finish in time.

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: bad6095b626d22def4d0c166597a972f842fa06e
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-28d37e/5154/summary.html
CMSSW: CMSSW_11_1_X_2020-03-11-2300
SCRAM_ARCH: slc7_amd64_gcc820

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-28d37e/5154/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 34
 - DQMHistoTests: Total histograms compared: 2680593
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2680273
 - DQMHistoTests: Total skipped: 319
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 33 files compared)
 - Checked 147 log files, 16 edm output root files, 34 DQM output files

schneiml commented:
+1

Seems to work in the PR tests now, runs only a single thread per test job, and does actually fail when there are legacy modules (I tested that by adding "HARVESTING" to the list of steps to check, which contains legacy modules).

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @davidlange6, @silviodonato, @fabiocos (and backports should be raised in the release meeting by the corresponding L2)

silviodonato commented:
+1

silviodonato commented:
@smuzaffar @mrodozov please note the next IB tests of CMSSW_11_1_X_2020-03-16-1100 will run the new unit test https://github.com/cms-sw/cmssw/blob/master/DQMOffline/Configuration/test/BuildFile.xml
